File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/rdd/CustomersNeverOrdered.scala
package io.devorbit.sparklabs.batch.sales.rdd

object CustomersNeverOrdered extends App {
  val orderedCustomers = RddUtils.orders.map(_.customer_id).distinct().collect().toSet
  RddUtils.customers
    .filter(customer => !orderedCustomers.contains(customer.customer_id))
    .collect()
    .foreach(println)
}

File: python/sparklabs/batch/sales/rdd/customers_never_ordered.py
from sparklabs.batch.sales.rdd import utils


def run() -> None:
    ordered_customers = set(utils.orders.map(lambda order: order["customer_id"]).distinct().collect())
    for row in utils.customers.filter(
        lambda customer: customer["customer_id"] not in ordered_customers
    ).collect():
        print(row)


if __name__ == "__main__":
    run()
